### ISSUE TITLE
Estimated comptetitive price: unavailable

### ISSUE THREAD

**Issue by david60** [NebulousLabs/Sia#1315]
Sorry to add a ticket in the midst of finalising your release; I report it now just to give you what information I have. I've seen the estimated price become unavailable in a recent master build. I suppose this is probably #1281 and/or #1284 (however the fix for the latter was already included in my build). I've been considering this change:

```go
diff --git a/modules/renter/hostdb/scan.go b/modules/renter/hostdb/scan.go
index 765238f..a576439 100644
--- a/modules/renter/hostdb/scan.go
+++ b/modules/renter/hostdb/scan.go
@@ -132,6 +132,14 @@ func (hdb *HostDB) threadedProbeHosts() {
                                return
                        }
 
+                       // If there's already a host of the same id, remove that host
+                       // so that it may be readded with changed weight
+                       priorEntry, exists := hdb.activeHosts[hostEntry.NetAddress]
+                       if exists {
+                               priorEntry.removeNode()
+                               delete(hdb.activeHosts, hostEntry.NetAddress)
+                       }
+
                        // Update the host settings, reliability, and weight. The old NetAddress
                        // must be preserved.
                        settings.NetAddress = hostEntry.HostExternalSettings.NetAddress
                        hostEntry.HostExternalSettings = settings
                        hostEntry.Reliability = MaxReliability
                        hostEntry.Weight = calculateHostWeight(*hostEntry)
                        hostEntry.Online = true

                        // If 'maxActiveHosts' has not been reached, add the host to the
                        // activeHosts tree.
                        if _, exists := hdb.activeHosts[hostEntry.NetAddress]; exists || len(hdb.activeHosts) < maxActiveHosts {
                                hdb.insertNode(hostEntry)
                        }
```

(Assumption is that during hdb.insertNode(hostEntry), if the entry already exists removeNode() is called internally first. But as this is after the weight has been modified the adjustment of the weight sum is wrong. So I was proposing to remove the entry before it is changed).

I would have reported this as a test to reproduce a problem, but given the release schedule maybe it's better just to give you the chance to look at it now.

Also, typo "comptetitive".

**Comment by DavidVorick** [comment]
OH. You are correct, we update the weight before we remove the node in the `insertNode` action, which causes corruption when `removeNode` is called. I was looking for the reason earlier today.

In general, this is pretty poorly designed code. Ideally, we would not be able to change the state of the hostEntry while it is still in the tree, as the correctness properties of the tree depend on the state of the hostEntry being constant while in-use. I have plans to write a safer hostdb for 1.2.

Thanks again, I owe you at least a beer or something. You've caught at least 3 fairly significant issues and helped us resolve them.

**Comment by DavidVorick** [comment]
I do not see the typo you mentioned

**Comment by david60** [comment]
I found it in two places in siac/hostcmd.go (actually with a capital, "Comptetitive")

**Comment by DavidVorick** [comment]
Ok, barring a failed code review, I believe this issue is resolved in #1317
